File: src/main/java/org/parthvnp/Graph/TopologicalSort.java
package org.parthvnp.Graph;

import java.util.*;

public class TopologicalSort {
  public HashSet<Integer> visited = new HashSet<>();
  public Queue<Integer> ans = new ArrayDeque<>();

  public int[] Sort(int n, int[][] edges) {
    var map = construct(edges);
    for (int i = 0; i < n; i++) {
      if(!visited.contains(i)) dfs(i, map);
    }
    while(!ans.isEmpty()){
      System.out.println(ans.poll());
    }
    return new int[2];
  }

  public static HashMap<Integer, ArrayList<Integer>> construct(int[][] edges) {
    var map = new HashMap<Integer, ArrayList<Integer>>();
    for (int[] i : edges) {
      map.computeIfAbsent(i[1], k -> new ArrayList<>());
      map.computeIfPresent(
          i[1],
          (k, v) -> {
            v.add(i[0]);
            return v;
          });
    }
    return map;
  }

  public void dfs(int vertex, HashMap<Integer, ArrayList<Integer>> graph) {
    visited.add(vertex);
    for (int n : graph.getOrDefault(vertex, new ArrayList<>())) {
      if (!visited.contains(n)) dfs(n, graph);
    }
    ans.offer(vertex);
  }

  public static void main(String[] args) {
    var s = new TopologicalSort();
    s.Sort(5, new int[][]{new int[]{0,1},  new int[]{1,2}, new int[] {2,3}, new int[] {3,4}});
  }
}

File: src/main/python/graph/topological_sort.py
from typing import *

class TopologicalSort:
    def sort(self, n: int, edges: List[List[int]]):
        def construct():
            adj = {v: [] for v in range(n)}
            for f, j in edges: adj[j].append(f)
            return adj
        graph = construct()
        visited  = set()
        ans = []
        def dfs(vertex: int):
            visited.add(vertex)
            for k in graph[vertex]:
                if k not in visited: dfs(k)
            ans.append(vertex)
        for i in range(n):
            if i not in visited: dfs(i)
        return ans


if __name__ == '__main__':
    s = TopologicalSort()
    r = s.sort(2, [[1, 0], [0,1]])
    print(r)
